base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,811 @@
+            for (zerr = UNZ_OK; zerr == UNZ_OK;)
+            {
+                len = zerr = unzReadCurrentFile(hzf, buf, len);
+                if (zerr <= 0)
+                    break;
+                zerr = WriteFile(hFile, buf, len, &cb, NULL) && cb == len ? UNZ_OK : UNZ_ERRNO;
+            }
+            CloseHandle(hFile);
+        }
+        unzCloseCurrentFile(hzf);
+    }
+    unzClose(hzf);
+    return zerr == UNZ_END_OF_LIST_OF_FILE ? UNZ_OK : zerr;
+}
+
+static UINT
+ExtractZip(LPCWSTR Archive, const CStringW &OutputDir,
+           EXTRACTCALLBACK Callback, void *Cookie)
+{
+    int zerr = ExtractFilesFromZip(Archive, OutputDir, Callback, Cookie);
+    return zerr == UNZ_ERRNO ? GetLastError() : zerr ? ERROR_INTERNAL_ERROR : 0;
+}
+
+static UINT
+ExtractCab(LPCWSTR Archive, const CStringW &OutputDir,
+           EXTRACTCALLBACK Callback, void *Cookie)
+{
+    if (ExtractFilesFromCab(Archive, OutputDir, Callback, Cookie))
+        return ERROR_SUCCESS;
+    UINT err = GetLastError();
+    return err ? err : ERROR_INTERNAL_ERROR;
+}
+
+enum { IM_STARTPROGRESS = WM_APP, IM_PROGRESS, IM_END };
+
+static struct CommonInfo
+{
+    LPCWSTR AppName;
+    HWND hDlg;
+    DWORD Error, Count;
+    BOOL Silent;
+    CRegKey *ArpKey, Entries;
+
+    CommonInfo(LPCWSTR DisplayName, BOOL IsSilent = FALSE)
+    : AppName(DisplayName), hDlg(NULL), Error(0), Count(0), Silent(IsSilent)
+    {
+    }
+    inline HWND GetGuiOwner() const { return Silent ? NULL : hDlg; }
+} *g_pInfo;
+
+struct InstallInfo : CommonInfo
+{
+    CConfigParser &Parser;
+    LPCWSTR ArchivePath, InstallDir, ShortcutFile;
+    CStringW UninstFile, MainApp;
+    UINT InstallDirLen, EntryCount;
+    BOOL PerUser;
+
+    InstallInfo(LPCWSTR AppName, CConfigParser &CP, LPCWSTR Archive)
+    : CommonInfo(AppName), Parser(CP), ArchivePath(Archive)
+    {
+        EntryCount = 0;
+    }
+};
+
+static UINT
+ErrorBox(UINT Error = GetLastError())
+{
+    if (!Error)
+        Error = ERROR_INTERNAL_ERROR;
+    WCHAR buf[400];
+    UINT fmf = FORMAT_MESSAGE_IGNORE_INSERTS | FORMAT_MESSAGE_FROM_SYSTEM;
+    FormatMessageW(fmf, NULL, Error, 0, buf, _countof(buf), NULL);
+    MessageBoxW(g_pInfo->GetGuiOwner(), buf, 0, MB_OK | MB_ICONSTOP);
+    g_pInfo->Error = Error;
+    return Error;
+}
+
+static LPCWSTR
+GetCommonString(LPCWSTR Name, CStringW &Output, LPCWSTR Default = L"")
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    BOOL found = Info.Parser.GetString(Name, Output);
+    return (found && !Output.IsEmpty() ? Output : Output = Default).GetString();
+}
+
+static LPCWSTR
+GetGenerateString(LPCWSTR Name, CStringW &Output, LPCWSTR Default = L"")
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    UINT r = Info.Parser.GetSectionString(DB_GENINSTSECTION, Name, Output);
+    return (r ? Output : Output = Default).GetString();
+}
+
+static void
+WriteArpEntry(LPCWSTR Name, LPCWSTR Value, UINT Type = REG_SZ)
```

Comment:
What's Arp?

Reply:
"Add/remove programs" [a common name](https://learn.microsoft.com/en-us/windows/win32/msi/configuring-add-remove-programs-with-windows-installer) for these values

Reply:
Add comment for that.